src/algorithms/advi.jl

```diff
@@ -0,0 +1,82 @@
+"""
+$(TYPEDEF)
+
+Automatic Differentiation Variational Inference (ADVI) with automatic differentiation
+backend `AD`.
+
+# Fields
+
+$(TYPEDFIELDS)
+"""
+struct ADVI{AD} <: VariationalInference{AD}
+    "Number of samples used to estimate the ELBO in each optimization step."
+    samples_per_step::Int
+    "Maximum number of gradient steps."
+    max_iters::Int
+end
+
+function ADVI(samples_per_step::Int = 1, max_iters::Int = 1000)
+    return ADVI{ADBackend()}(samples_per_step, max_iters)
+end
+
+alg_str(::ADVI) = "ADVI"
+nsamples(alg::ADVI) = alg.samples_per_step
+niters(alg::ADVI) = alg.max_iters
+
+function compats(::ADVI)
+    return Union{
+                CholMvNormal,
+                Bijectors.TransformedDistribution{<:CholMvNormal},
+                DiagMvNormal,
+                Bijectors.TransformedDistribution{<:DiagMvNormal},
+            }
+end
+
+function init(alg::ADVI, q, opt)
+    samples_per_step = nsamples(alg)
+    x₀ = rand(q, samples_per_step) # Preallocating x₀
+    x = similar(x₀) # Preallocating x
+    diff_result = DiffResults.GradientResult(x)
+    return (x₀=x₀, x=x, diff_result=diff_result)
+end
+
+function step!(::ELBO, alg::ADVI, q, logπ, state, opt)
+    randn!(state.x₀) # Get initial samples from x₀
+    reparametrize!(state.x, q, state.x₀)
+    f(X) = sum(eachcol(X)) do x
+        return evaluate(logπ, q, x)
+    end
+    grad!(state.diff_result, f, state.x, alg)
+    return update!(alg, q, state, opt)
+end
+
+function update!(alg::ADVI, q, state, opt)
+    Δ = DiffResults.gradient(state.diff_result)
+    update_mean!(q, vec(mean(Δ, dims = 2)), opt)
+    update_cov!(alg, q, Δ, state, opt)
```

Comment:
Well actually not, and that's where I am trying to be maybe more precise. The ADVI approach I implemented is following the ADVI paper which namely states : 
![image](https://user-images.githubusercontent.com/11480067/114428108-15fd2280-9bbc-11eb-87ca-020bb409b9e4.png)
So the underlying distribution is always a Gaussian.
I will add a reference to the paper in the docs

If we want to train something different it will have to be a different algorithm.

Reply:
I'm fully aware that the original paper only had this in mind, but it's unecessarily restricted and is mainly just a consequence of the restricted tools they had available at the time.

It's called "Automatic Differential Variational Inference" which really refers to the fact that we use AD + reparameterization trick.

IMO it just comes down to: what do it cost us to allow *any* reparameterization with *any* valid base distribution, and just let Gaussian be the default? That just seems overall superior, no?

Reply:
Well in theory yes that sounds amazing. The problem comes from how do you formulate the reparametrization trick for non-Gaussian. It's not always possible right?

Reply:
Normalizing flows; affine transformation is just an instance:)

Reply:
IMO, we don't even need all these implementations of differnt Gaussians. We just make it different affine transformations, e.g. introduce a `Affine <: Bijector` and define different evaluations depending on whether the scaling-matrix is `Cholesky`, etc.

Reply:
The above is the reasoning behind the current impl btw. Though I 100% agree we should make it more convenient to instantiate the different approaches.